File: leetcode/053-Maximum-Subarray/tina.swift
class Solution {
    func maxSubArray(_ nums: [Int]) -> Int {
        // SUM(j...k) = SUM(0...k) - SUM(0...j - 1)
        // ^ keep max   ^ current    ^ keep min
        var max_sub_sum = nums[0]
        var min_sub_sum = 0
        var sum = 0
        
        for num in nums {
            sum += num
            
            if sum - min_sub_sum > max_sub_sum {
                max_sub_sum = sum - min_sub_sum
            }
            
            if sum < min_sub_sum {
                min_sub_sum = sum
            }
        }
        
        return max_sub_sum
    }
}

File: leetcode/223-Rectangle-Area/tina.cpp
class Solution {
public:
    int computeArea(int A, int B, int C, int D, int E, int F, int G, int H) {
        int area = (C - A) * (D - B) + (G - E) * (H - F);
        int l = max(A, E);
        int r = min(C, G);
        int t = min(D, H);
        int b = max(B, F);
        
        if (l < r && b < t) {
            area -= (r - l) * (t - b);
        }
        
        return area;
    }
};
